substrate/primitives/genesis-builder/src/lib.rs

```diff
@@ -19,39 +19,92 @@
+//! `RuntimeGenesisConfig` structure in a form of the JSON blob.
 //!
-//! Providing externalities with empty storage and putting `RuntimeGenesisConfig` into storage
-//! allows to catch and build the raw storage of `RuntimeGenesisConfig` which is the foundation for
-//! genesis block.
+//! Additionally the runtime may provide a number of partial predefined `RuntimeGenesisConfig`
+//! configurations in the form of patches which shall be applied on top of the default
+//! `RuntimeGenesisConfig`. The patch is a JSON blob, which essentially comprises the list of
+//! key-value pairs that are to be customized in the default runtime genesis config.
+//! These predefined configurations are referred to as presets.
+//!
+//! This allows the runtime to provide a number of predefined configs (e.g. for different
+//! testnets or development) without neccessity to leak the runtime types outside the itself (e.g.
+//! node or chain-spec related tools).
+//!
+//! This Runtime API allows to interact with `RuntimeGenesisConfig`, in particular:
+//! - provide the list of available preset names,
+//! - provide a number of named presets of `RuntimeGenesisConfig`,
+//! - provide a JSON represention of the default `RuntimeGenesisConfig` (by simply serializing the
+//!   default `RuntimeGenesisConfig` struct into JSON format),
+//! - deserialize the full `RuntimeGenesisConfig` from given JSON blob and put the resulting
+//!   `RuntimeGenesisConfig` structure into the state storage creating the initial runtime's state.
+//!   Allows to build customized genesis. This operation internally calls `GenesisBuild::build`
+//!   function for all runtime pallets.
+//!
+//! Providing externalities with an empty storage and putting `RuntimeGenesisConfig` into storage
+//! (by calling `build_state`) allows to construct the raw storage of `RuntimeGenesisConfig`
+//! which is the foundation for genesis block.
 
 extern crate alloc;
+use alloc::vec::Vec;
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
 
 /// The result type alias, used in build methods. `Err` contains formatted error message.
 pub type Result = core::result::Result<(), sp_runtime::RuntimeString>;
 
+/// Wrapper representing preset id.
+#[derive(Decode, Encode, Debug, TypeInfo, PartialEq)]
+pub struct PresetId(Vec<u8>);
+
+impl From<&str> for PresetId {
+	fn from(v: &str) -> Self {
+		Self(v.as_bytes().to_vec())
+	}
+}
+
+impl<'a> TryFrom<&'a PresetId> for &'a str {
+	type Error = core::str::Utf8Error;
+	fn try_from(v: &'a PresetId) -> core::result::Result<&'a str, Self::Error> {
+		core::str::from_utf8(&v.0)
+	}
+}
+
 sp_api::decl_runtime_apis! {
 	/// API to interact with RuntimeGenesisConfig for the runtime
 	pub trait GenesisBuilder {
-		/// Creates the default `RuntimeGenesisConfig` and returns it as a JSON blob.
+		/// Build `RuntimeGenesisConfig` from a JSON blob not using any defaults and store it in the
```

Comment:
For example in a NON-frame based runtime, you can still use this API, and you receive this JSON, and you somehow put it in the storage, and there is no `RuntimeGenesisConfig` involved.

Reply:
I added some note for non-FRAME runtimes.

Reply:
92e6824b1a2